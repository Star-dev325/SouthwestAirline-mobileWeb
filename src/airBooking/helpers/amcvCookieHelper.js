// @flow
import _ from 'lodash';
import * as Cookie from 'src/shared/swa-persistence/cookie';

import { buildPathWithQuery } from 'src/shared/helpers/pathUtils';

const AMCV_COOKIE_NAME = 'AMCV_65D316D751E563EC0A490D4C%40AdobeOrg';
const AMCV_COOKIE_KEY_SEPARATOR = 'MID';
const AMCV_COOKIE_PIPE_SEPARATOR = '|';
const CHASE_URL_KEYWORD = 'creditcard';

export type ChaseAppendParams = {
  isChaseCombo?: boolean,
  pageId?: string
};

export const getMcvid = () => {
  const value = Cookie.getValue(AMCV_COOKIE_NAME);

  return _.chain(value).split(AMCV_COOKIE_KEY_SEPARATOR).nth(1).split(AMCV_COOKIE_PIPE_SEPARATOR).nth(1).value() || '';
};

// Deprecated - please use appendParamsIfChaseUrl instead
export const addMvcidToChaseUrl = (url: string, isChase: boolean) => {
  let updateUrl = url;

  if (!_.isEmpty(url) && (isChase || url.includes(CHASE_URL_KEYWORD))) {
    const mcvid = getMcvid();

    if (!_.isEmpty(mcvid)) {
      updateUrl = `${url}&mcvid=${mcvid}`;
    }
  }

  return updateUrl;
};

export const appendParamsIfChaseUrl = (target: string = '', params: ChaseAppendParams = {}) => {
  const { isChaseCombo = false, pageId = '' } = params;
  const query = {};
  const isChaseUrl = _.includes(target, CHASE_URL_KEYWORD) || isChaseCombo;
  const mcvid = isChaseUrl && getMcvid();

  mcvid && _.set(query, 'mcvid', mcvid);
  pageId && _.set(query, 'pageId', pageId);

  return buildPathWithQuery(target, query);
};
