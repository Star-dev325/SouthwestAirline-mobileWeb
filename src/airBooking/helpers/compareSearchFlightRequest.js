import _ from 'lodash';
import { transformToAPIRequest } from 'src/shared/transformers/flightProductSearchRequestTransformer';

export const compareSearchFlightRequest = (requestOne, requestTwo) => {
  const transformedRequestOneQuery = _.get(transformToAPIRequest(requestOne), 'query');
  const transformedRequestTwoQuery = _.get(transformToAPIRequest(requestTwo), 'query');
  const requestOneWithoutPromoCode = _.omit(transformedRequestOneQuery, 'promo-code');
  const requestTwoWithoutPromoCode = _.omit(transformedRequestTwoQuery, 'promo-code');

  return _.isEqual(requestOneWithoutPromoCode, requestTwoWithoutPromoCode);
};
