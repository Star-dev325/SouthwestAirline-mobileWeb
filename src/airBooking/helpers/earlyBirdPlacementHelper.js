// @flow

import _ from 'lodash';
import numeral from 'numeral';
import i18n from '@swa-ui/locale';
import type { EarlyBirdEligibility } from 'src/airBooking/flow-typed/airBooking.types';
import type { CurrencyType } from 'src/shared/flow-typed/shared.types';

const getEarlyBirdPrefixText = (unitPriceInBound: CurrencyType, unitPriceOutBound: CurrencyType) => {
  const showStartingFromMessage =
    !!unitPriceOutBound &&
    !!unitPriceInBound &&
    _.get(unitPriceOutBound, 'amount') !== _.get(unitPriceInBound, 'amount');

  return showStartingFromMessage ? i18n('SHARED__EARLY_BIRD__PER_PASSENGER_PRICE_MESSAGE_PREFIX') : '(';
};

const getEarlyBirdUnitAmount = (unitPriceInBound: CurrencyType, unitPriceOutBound: CurrencyType) => {
  const earlyBirdUnitPrice = _.chain([unitPriceOutBound, unitPriceInBound])
    .compact()
    .minBy((item) => numeral(item.amount).value())
    .value();

  return _.get(earlyBirdUnitPrice, 'amount');
};

export const getEarlyBirdAdditionalTemplateData = (
  unitPriceInBound: CurrencyType,
  unitPriceOutBound: CurrencyType,
  earlyBirdEligibility: ?EarlyBirdEligibility
) => ({
  earlyBirdTotalPrice: _.get(earlyBirdEligibility, 'totalPrice.amount'),
  earlyBirdUnitPrice: getEarlyBirdUnitAmount(unitPriceInBound, unitPriceOutBound),
  earlyBirdPrefixText: getEarlyBirdPrefixText(unitPriceInBound, unitPriceOutBound)
});
