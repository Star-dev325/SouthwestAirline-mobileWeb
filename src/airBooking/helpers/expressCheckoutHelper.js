// @flow
import _ from 'lodash';
import type { Passenger } from 'src/shared/flow-typed/shared.types';
import validator from 'src/shared/form/formValidators/validator';

export const isInformationCompletedForExpressCheckout = (passengerDetailsPage: Passenger): boolean => {
  const { firstName, lastName, dateOfBirth, gender } = passengerDetailsPage || {};

  return !_.isEmpty(firstName) && !_.isEmpty(lastName) && !_.isEmpty(dateOfBirth) && validator.isValidGender(gender);
};
