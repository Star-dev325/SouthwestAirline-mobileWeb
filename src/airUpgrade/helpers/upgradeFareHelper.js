// @flow
import type { UpgradeFareReservationDataType } from 'src/shared/flow-typed/shared.types';
import type { PricingDataType, BoundSelectionDataType } from 'src/airUpgrade/flow-typed/airUpgrade.types';

export const buildUpgradeFareReservationRequest = ({
  firstName,
  lastName,
  recordLocator,
  passengerSearchToken,
  nonPremiumSearch,
  link
}: UpgradeFareReservationDataType) =>
  (link?.body?.passengerSearchToken
    ? link
    : {
      href: `/v1/mobile-air-booking/page/upgrade/${recordLocator}`,
      body: {
        firstName,
        lastName,
        recordLocator,
        passengerSearchToken,
        nonPremiumSearch
      },
      method: 'POST'
    });

export const getPricingChangeRequests = (
  changePricingLink: Link,
  pricingDataList: Array<PricingDataType>,
  boundSelectionDataList: Array<BoundSelectionDataType>
) => {
  const changeRequests = changePricingLink.body?.changeRequests ?? [];
  const fundsAppliedToken = changePricingLink.body?.fundsAppliedToken;

  const changeRequestsWithProductId = boundSelectionDataList.map<BoundSelectionDataType>(
    (boundSelectionDataType, index): BoundSelectionDataType => {
      const upgradeChangeRequest = changeRequests[index];

      if (boundSelectionDataType.canUpgrade) {
        const isSelected = pricingDataList.some(
          (pricingData) => pricingData.productId === boundSelectionDataType.productId && pricingData.isSelected
        );

        return isSelected
          ? {
            ...upgradeChangeRequest,
            productId: boundSelectionDataType.productId,
            arrivalAirportCode: boundSelectionDataType.arrivalAirportCode,
            departureAirportCode: boundSelectionDataType.departureAirportCode
          }
          : { ...upgradeChangeRequest };
      }

      return { ...upgradeChangeRequest };
    }
  );

  return { changeRequests: changeRequestsWithProductId, fundsAppliedToken };
};
