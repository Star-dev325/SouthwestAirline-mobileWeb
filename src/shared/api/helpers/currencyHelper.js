import _ from 'lodash';
import numeral from 'numeral';

export const getCurrencyFromDollarsToCents = (price) => {
  const dollars = Number(price.value);

  return _.isNaN(dollars) ? price.value : parseFloat((dollars * 100).toFixed(2));
};

export const addCurrency = (...arrayOfCurrencyType) => {
  const prices = _.filter(arrayOfCurrencyType, (price) => !_.isEmpty(price));

  const currencyInfo = _.pick(prices[0], ['currencyCode', 'currencySymbol']);
  const isSameCurrency = _.every(prices, (price) => price.currencyCode === currencyInfo.currencyCode);

  if (!isSameCurrency) return;

  const amount = _.reduce(prices, (total, price) => numeral(_.get(price, 'amount')).add(total).value(), 0);

  return _.merge({}, currencyInfo, {
    amount: numeral(amount).format('0,000.00')
  });
};
