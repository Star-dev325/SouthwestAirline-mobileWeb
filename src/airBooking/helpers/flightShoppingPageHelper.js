import _ from 'lodash';
import { permutations } from 'src/shared/helpers/mathUtils';
import { INBOUND, OUTBOUND } from 'src/shared/constants/flightBoundTypes';
import PassengerTypes from 'src/shared/constants/passengerTypes';

const { ADULT } = PassengerTypes;

export const getNextFlightShoppingPageParams = (params, response) => {
  const { inboundPage } = response;
  const bounds = inboundPage ? [OUTBOUND, INBOUND] : [OUTBOUND];
  const paxTypes = [ADULT];

  const pages = permutations(paxTypes, bounds).map((paxAndBound) => ({
    paxType: paxAndBound[0],
    direction: paxAndBound[1]
  }));

  const index = _.findIndex(pages, params);

  return index === -1 ? undefined : _.get(pages, `[${index + 1}]`);
};

export const getFirstShoppingPageParams = () => ({
  paxType: ADULT,
  direction: OUTBOUND
});

export const generateFlightShoppingPages = (response) => {
  const { outboundPage, inboundPage } = response.flightShoppingPage;
  const isRoundTrip = !!inboundPage;
  let pages = [];

  const buildPages = () => {
    const result = [
      {
        direction: OUTBOUND,
        paxType: ADULT,
        ...outboundPage
      }
    ];

    if (isRoundTrip) {
      result.push({
        direction: INBOUND,
        paxType: ADULT,
        ...inboundPage
      });
    }

    return result;
  };

  pages = pages.concat(buildPages());

  return pages;
};
