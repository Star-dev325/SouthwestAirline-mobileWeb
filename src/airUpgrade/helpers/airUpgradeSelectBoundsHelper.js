import { addCurrency } from 'src/shared/api/helpers/currencyHelper';
import { toNumberFromFormattedString, toFormattedStringFromNumber } from 'src/shared/helpers/currencyValueHelper';
import { POINTS } from 'src/shared/constants/currencyTypes';

export const flattenUpgradeBounds = (upgradeBounds) =>
  upgradeBounds.reduce((flattenedBounds, upgradeBound) => {
    const { boundSelectionDataList = [], upgradeablePricingDataList = [] } = flattenedBounds;

    const {
      arrivalAirportCode,
      arrivalTime,
      boundType,
      departureAirportCode,
      departureDate,
      departureDayOfWeek,
      departureTime,
      flight,
      isOvernight,
      isNextDayArrival,
      numberOfPassengers,
      upgradeFare: {
        upgradePrice = null,
        upgradeTotalPrice = null,
        upgradeMessage: { header: upgradeMessageHeader, body: upgradeMessageBody },
        _meta: { productId, canUpgrade }
      }
    } = upgradeBound;

    const boundSelectionData = {
      arrivalAirportCode,
      arrivalTime,
      boundType,
      canUpgrade,
      departureAirportCode,
      departureDate,
      departureDayOfWeek,
      departureTime,
      isOvernight,
      isNextDayArrival,
      productId,
      upgradeMessageBody,
      upgradeMessageHeader
    };

    const pricingData = {
      arrivalAirportCode,
      boundType,
      canUpgrade,
      departureAirportCode,
      flight,
      isSelected: false,
      numberOfPassengers,
      productId,
      upgradePrice,
      upgradeTotalPrice
    };

    return {
      boundSelectionDataList: [...boundSelectionDataList, boundSelectionData],
      upgradeablePricingDataList: canUpgrade ? [...upgradeablePricingDataList, pricingData] : upgradeablePricingDataList
    };
  }, {});

export const updateSelectedPricingData = (pricingDataList, { productId, isSelected }) =>
  pricingDataList.map((pricingData) =>
    (pricingData.productId === productId ? { ...pricingData, isSelected } : pricingData)
  );

export const addPoints = (...arrayOfCurrencyType) => {
  const { currencyCode = POINTS, currencySymbol = '', amount: unformattedAmount } = addCurrency(...arrayOfCurrencyType);
  const amountAsNumber = toNumberFromFormattedString(unformattedAmount);
  const amount = toFormattedStringFromNumber(amountAsNumber);

  return {
    currencyCode,
    currencySymbol,
    amount
  };
};
