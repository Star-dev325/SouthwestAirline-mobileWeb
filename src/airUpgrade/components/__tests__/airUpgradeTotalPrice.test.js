import { render } from '@testing-library/react';
import React from 'react';
import AirUpgradeTotalPrice from 'src/airUpgrade/components/airUpgradeTotalPrice';
import * as airUpgradeSelectBoundsHelper from 'src/airUpgrade/helpers/airUpgradeSelectBoundsHelper';
import * as currencyHelper from 'src/shared/api/helpers/currencyHelper';
import AirUpgradeViewReservationApiJsonBuilder from 'test/builders/apiResponse/airUpgradeViewReservationApiJsonBuilder';

describe('AirUpgradeTotalPrice', () => {
  let mockAddCurrency;
  let mockAddPoints;

  beforeEach(() => {
    mockAddCurrency = jest.spyOn(currencyHelper, 'addCurrency').mockReturnValue({
      amount: '6,000.00',
      currencyCode: 'USD',
      currencySymbol: '$'
    });

    mockAddPoints = jest.spyOn(airUpgradeSelectBoundsHelper, 'addPoints').mockReturnValue({
      amount: '120,000',
      currencyCode: 'PTS'
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  const getPointsPricingDataList = () =>
    new AirUpgradeViewReservationApiJsonBuilder().withPointsBooking().build().viewUpgradeReservationPage
      .pricingDataList;
  const getPointsBooking = () =>
    new AirUpgradeViewReservationApiJsonBuilder().withPointsBooking().withPointsBothBoundsNotSelected().build()
      .viewUpgradeReservationPage.pricingDataList;
  const getPricingDataList = () =>
    new AirUpgradeViewReservationApiJsonBuilder().build().viewUpgradeReservationPage.pricingDataList;

  it('Should render a default amount of 0.00 if no bounds have been selected', () => {
    const pricingDataList = getPricingDataList();
    const { container } = createComponent(pricingDataList);

    expect(container).toMatchSnapshot();
  });

  it('Should render the total amount of one selected bound', () => {
    const [bound1, bound2] = getPricingDataList();
    const selectedBound = { ...bound1, isSelected: true };
    const { container } = createComponent([selectedBound, bound2]);

    expect(container).toMatchSnapshot();
  });

  it('Should render the sum of all dollar bounds if they are selected', () => {
    const { container } = createComponent(getPricingDataList().map((bound) => ({ ...bound, isSelected: true })));

    expect(container).toMatchSnapshot();
  });

  it('Should call addCurrency when all dollar bounds are selected', () => {
    createComponent(getPricingDataList().map((bound) => ({ ...bound, isSelected: true })));

    expect(mockAddCurrency).toHaveBeenCalled();
  });

  it('Should not call addPoints when all dollar bounds are selected', () => {
    createComponent(getPricingDataList().map((bound) => ({ ...bound, isSelected: true })));

    expect(mockAddPoints).not.toHaveBeenCalled();
  });

  it('Should render points booking total correctly with 1 selected bound', () => {
    const [bound1, bound2] = getPointsPricingDataList();
    const selectedBound = {
      ...bound1,
      isSelected: true
    };
    const { container } = createComponent([selectedBound, bound2]);

    expect(container).toMatchSnapshot();
  });

  it('Should render points booking total correctly with no selected bound', () => {
    const { container } = createComponent(
      getPointsBooking().map((bound) => ({ ...bound, isSelected: false })),
      false
    );

    expect(container).toMatchSnapshot();
  });

  it('Should render points booking total correctly with 2 selected bounds', () => {
    const { container } = createComponent(getPointsPricingDataList().map((bound) => ({ ...bound, isSelected: true })));

    expect(container).toMatchSnapshot();
  });
  it('Should call addPoints with 2 selected point bounds', () => {
    createComponent(getPointsPricingDataList().map((bound) => ({ ...bound, isSelected: true })));

    expect(mockAddPoints).toHaveBeenCalled();
  });

  it('Should not call addCurrency with 2 selected point bounds', () => {
    createComponent(getPointsPricingDataList().map((bound) => ({ ...bound, isSelected: true })));

    expect(mockAddCurrency).not.toHaveBeenCalled();
  });

  const createComponent = (pricingDataList) => render(<AirUpgradeTotalPrice pricingDataList={pricingDataList} />);
});
