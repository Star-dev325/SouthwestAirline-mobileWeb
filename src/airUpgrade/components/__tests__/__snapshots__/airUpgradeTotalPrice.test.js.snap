// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirUpgradeTotalPrice Should render a default amount of 0.00 if no bounds have been selected 1`] = `
<div>
  <div
    class="price-total"
  >
    <div
      class="price-total--info"
    >
      <div
        class="price-line"
      >
        <span
          class="price-line--title"
        >
          SHARED__PRICE_LINE_TITLES__TOTAL_PASSENGER
        </span>
        <div
          class="price-amount"
        >
          <div
            class="price-line--currency_total"
          >
            <div
              class="currency"
            >
              <span
                class="money-sign"
              >
                $
              </span>
              <span
                data-qa="total-amount"
              >
                0.00
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="table fullwidth mt3"
      >
        <div
          class="price-total--info-col"
        >
          <p>
            SHARED__PRICE_TOTAL_EXPLANATORY_NOTE__BRIEF_NOTE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AirUpgradeTotalPrice Should render points booking total correctly with 1 selected bound 1`] = `
<div>
  <div
    class="price-total"
  >
    <div
      class="price-total--info"
    >
      <div
        class="price-line mb4"
      >
        <span
          class="price-line--title"
        >
          SHARED__PRICE_LINE_TITLES__POINTS_TOTAL
        </span>
        <div
          class="price-amount"
        >
          <div
            class="price-line--currency_total"
          >
            <div
              class="currency"
            >
              <span
                data-qa="total-amount"
              >
                120,000
              </span>
              <span
                class="points-sign"
              >
                SHARED__COMMON__POINTS_TEXT
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="table fullwidth mt3"
      >
        <div
          class="price-total--info-col"
        >
          <p>
            SHARED__PRICE_TOTAL_EXPLANATORY_NOTE__BRIEF_NOTE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AirUpgradeTotalPrice Should render points booking total correctly with 2 selected bounds 1`] = `
<div>
  <div
    class="price-total"
  >
    <div
      class="price-total--info"
    >
      <div
        class="price-line mb4"
      >
        <span
          class="price-line--title"
        >
          SHARED__PRICE_LINE_TITLES__POINTS_TOTAL
        </span>
        <div
          class="price-amount"
        >
          <div
            class="price-line--currency_total"
          >
            <div
              class="currency"
            >
              <span
                data-qa="total-amount"
              >
                120,000
              </span>
              <span
                class="points-sign"
              >
                SHARED__COMMON__POINTS_TEXT
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="table fullwidth mt3"
      >
        <div
          class="price-total--info-col"
        >
          <p>
            SHARED__PRICE_TOTAL_EXPLANATORY_NOTE__BRIEF_NOTE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AirUpgradeTotalPrice Should render points booking total correctly with no selected bound 1`] = `
<div>
  <div
    class="price-total"
  >
    <div
      class="price-total--info"
    >
      <div
        class="price-line mb4"
      >
        <span
          class="price-line--title"
        >
          SHARED__PRICE_LINE_TITLES__POINTS_TOTAL
        </span>
        <div
          class="price-amount"
        >
          <div
            class="price-line--currency_total"
          >
            <div
              class="currency"
            >
              <span
                data-qa="total-amount"
              >
                0
              </span>
              <span
                class="points-sign"
              >
                SHARED__COMMON__POINTS_TEXT
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="table fullwidth mt3"
      >
        <div
          class="price-total--info-col"
        >
          <p>
            SHARED__PRICE_TOTAL_EXPLANATORY_NOTE__BRIEF_NOTE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AirUpgradeTotalPrice Should render the sum of all dollar bounds if they are selected 1`] = `
<div>
  <div
    class="price-total"
  >
    <div
      class="price-total--info"
    >
      <div
        class="price-line"
      >
        <span
          class="price-line--title"
        >
          SHARED__PRICE_LINE_TITLES__TOTAL_PASSENGER
        </span>
        <div
          class="price-amount"
        >
          <div
            class="price-line--currency_total"
          >
            <div
              class="currency"
            >
              <span
                class="money-sign"
              >
                $
              </span>
              <span
                data-qa="total-amount"
              >
                6,000.00
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="table fullwidth mt3"
      >
        <div
          class="price-total--info-col"
        >
          <p>
            SHARED__PRICE_TOTAL_EXPLANATORY_NOTE__BRIEF_NOTE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AirUpgradeTotalPrice Should render the total amount of one selected bound 1`] = `
<div>
  <div
    class="price-total"
  >
    <div
      class="price-total--info"
    >
      <div
        class="price-line"
      >
        <span
          class="price-line--title"
        >
          SHARED__PRICE_LINE_TITLES__TOTAL_PASSENGER
        </span>
        <div
          class="price-amount"
        >
          <div
            class="price-line--currency_total"
          >
            <div
              class="currency"
            >
              <span
                class="money-sign"
              >
                $
              </span>
              <span
                data-qa="total-amount"
              >
                6,000.00
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="table fullwidth mt3"
      >
        <div
          class="price-total--info-col"
        >
          <p>
            SHARED__PRICE_TOTAL_EXPLANATORY_NOTE__BRIEF_NOTE
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
